pkg/services/live/features/watch.go

```diff
@@ -0,0 +1,205 @@
+package features
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+
+	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+
+	"github.com/grafana/authlib/types"
+	"github.com/grafana/grafana-app-sdk/logging"
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/live"
+
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/services/apiserver"
+	"github.com/grafana/grafana/pkg/services/live/model"
+)
+
+// WatchRunner will start a watch task and broadcast results
+type WatchRunner struct {
+	publisher      model.ChannelPublisher
+	configProvider apiserver.RestConfigProvider
+
+	watchingMu sync.Mutex
+	watching   map[string]*watcher
+}
+
+func NewWatchRunner(publisher model.ChannelPublisher, configProvider apiserver.RestConfigProvider) *WatchRunner {
+	return &WatchRunner{
+		publisher:      publisher,
+		configProvider: configProvider,
+		watching:       make(map[string]*watcher),
+	}
+}
+
+func (b *WatchRunner) GetHandlerForPath(_ string) (model.ChannelHandler, error) {
+	return b, nil // all dashboards share the same handler
+}
+
+// Valid paths look like: {version}/{resource}[={name}]/{user.uid}
+// * v0alpha1/dashboards/u12345
+// * v0alpha1/dashboards=ABCD/u12345
+func (b *WatchRunner) OnSubscribe(ctx context.Context, u identity.Requester, e model.SubscribeEvent) (model.SubscribeReply, backend.SubscribeStreamStatus, error) {
+	// To make sure we do not share resources across users, in clude the UID in the path
+	userID := u.GetIdentifier()
+	if userID == "" {
+		return model.SubscribeReply{}, backend.SubscribeStreamStatusPermissionDenied, fmt.Errorf("missing user identity")
+	}
+	if !strings.HasSuffix(e.Path, userID) {
+		return model.SubscribeReply{}, backend.SubscribeStreamStatusPermissionDenied, fmt.Errorf("path must end with user uid (%s)", userID)
```

Comment:
Hi Ryan! 👋  Just noticed comments in the previous PR. A bit out of context, but to clarify. It's also possible to build channel auth without using parts of channel string – can be request to DB, or JWT-based (centrifuge protocol allows subscribing to channels using per-subscription tokens). But of course, sth like this, based on channel string and userID is the most efficient and simple. And can be extended later anyway with any permission check on this level.

Reply:
hello!!!  In my ideal setup, we would have a single channel for each resource type, eg:
* dashboard.grafana.app/dashboards
* dashboard.grafana.app/librarypanels
* ...

and populate it with all events for these kinds.  Before sending each event, we would check it the user has access to that event (it based on the id in the payload)